<img src=".\images\Logo.png" width=150 align="left" /> <img src=".\images\Logo2.jpg" width=450 align="right" />


# <center><font color= #1e8449 > <b>CU45_Planificación y promoción del destino en base a los patrones en origen de los turistas</font></center>

<font color='steelblue'><b>Citizenlab Data Science Methodology > II - Data Processing Domain </font>
***
> # <font color='steelblue'> <b>05.- Data Collection</font>

<font color='steelblue'>Data Collection is the process to obtain and generate (if required) necessary data to model the problem.</font>

### 01. Obtener datos de geometrías municipios

* Obtener como sf de IGN con paquete mapSpain. Contornos de municipios, solo C. Madrid








<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Settings" data-toc-modified-id="Settings-1"><font color="green">Settings</font></a></span></li><li><span><a href="#Data-Load" data-toc-modified-id="Data-Load-2"><font color="green">Data Load</font></a></span></li><li><span><a href="#ETL-Processes" data-toc-modified-id="ETL-Processes-3">ETL Processes</a></span><ul class="toc-item"><li><span><a href="#Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files" data-toc-modified-id="Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files-3.1">Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files</a></span></li></ul></li><li><span><a href="#Synthetic-Data-Generation" data-toc-modified-id="Synthetic-Data-Generation-4">Synthetic Data Generation</a></span></li><li><span><a href="#Fake-Data-Generation" data-toc-modified-id="Fake-Data-Generation-5">Fake Data Generation</a></span></li><li><span><a href="#Open-Data" data-toc-modified-id="Open-Data-6">Open Data</a></span></li><li><span><a href="#Data-Save" data-toc-modified-id="Data-Save-7"><font color="green">Data Save</font></a></span></li><li><span><a href="#Main-Conclusions" data-toc-modified-id="Main-Conclusions-8"><font color=" #2874a6 ">Main Conclusions</font></a></span></li><li><span><a href="#Main-Actions" data-toc-modified-id="Main-Actions-9"><font color=" #2874a6 ">Main Actions</font></a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Acciones-done" data-toc-modified-id="Acciones-done-9.0.1"><font color=" #2874a6 ">Acciones done</font></a></span></li><li><span><a href="#Acctions-to-perform" data-toc-modified-id="Acctions-to-perform-9.0.2"><font color=" #2874a6 ">Acctions to perform</font></a></span></li></ul></li></ul></li></ul></div>

## <font color='green'>Settings</font>

### Encoding

Con la siguiente expresión se evitan problemas con el encoding al ejecutar el notebook. Es posible que deba ser eliminada o adaptada a la máquina en la que se ejecute el código.

In [1]:
Sys.setlocale(category = "LC_ALL", locale = "es_ES.UTF-8")

[1] "es_ES.UTF-8/es_ES.UTF-8/es_ES.UTF-8/C/es_ES.UTF-8/C"

### Packages to use

_ELIMINAR O AÑADIR LO QUE TOQUE. COPIAR VERSIONES AL FINAL Y QUITAR CÓDIGO DE VERSIONES_

- {tcltk} para selección interactiva de archivos locales
- {sf} para trabajar con georeferenciación
- {mapSpain} para obtener los contornos de municipios
- {dplyr} para explorar datos


In [3]:
library(sf)
library(mapSpain)
library(dplyr)

p <- c("tcltk", "sf", "mapSpain", "dplyr")


### Paths

In [4]:
iPath <- "Data/Input/"
oPath <- "Data/Output/"

## <font color='green'>Data Load</font>

No aplica al leer los datos en línea de fuentes abiertas.

## ETL Processes

### Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files

No aplica

## Synthetic Data Generation

No aplica

## Fake Data Generation

No aplica

## Open Data

* Obtención de los contornos con el paquete `mapSpain` de `rOpenSpain`

In [5]:
data <- esp_get_munic() |> 
  filter(cpro == "28") 


Estructura de  los datos:

In [6]:
data |> glimpse()

Rows: 179
Columns: 8
$ codauto       <chr> "13", "13", "13", "13", "13", "13", "13", "13", "13", "1…
$ ine.ccaa.name <chr> "Madrid, Comunidad de", "Madrid, Comunidad de", "Madrid,…
$ cpro          <chr> "28", "28", "28", "28", "28", "28", "28", "28", "28", "2…
$ ine.prov.name <chr> "Madrid", "Madrid", "Madrid", "Madrid", "Madrid", "Madri…
$ cmun          <chr> "001", "002", "003", "004", "005", "006", "007", "008", …
$ name          <chr> "Acebeda, La", "Ajalvir", "Alameda del Valle", "Álamo, E…
$ LAU_CODE      <chr> "28001", "28002", "28003", "28004", "28005", "28006", "2…
$ geometry      <GEOMETRY [°]> POLYGON ((-3.61592 41.06982..., POLYGON ((-3.51…


Muestra de datos:

In [7]:
data |> tibble() |> slice_head(n = 5)

codauto ine.ccaa.name        cpro ine.prov.name cmun name             
1 13      Madrid, Comunidad de 28   Madrid        001  Acebeda, La      
2 13      Madrid, Comunidad de 28   Madrid        002  Ajalvir          
3 13      Madrid, Comunidad de 28   Madrid        003  Alameda del Valle
4 13      Madrid, Comunidad de 28   Madrid        004  Álamo, El        
5 13      Madrid, Comunidad de 28   Madrid        005  Alcalá de Henares
  LAU_CODE geometry                      
1 28001    POLYGON ((-3.61592 41.06982...
2 28002    POLYGON ((-3.51206 40.5386,...
3 28003    POLYGON ((-3.89848 40.95879...
4 28004    POLYGON ((-4.02386 40.24893...
5 28005    POLYGON ((-3.42668 40.45315...

## ETL Processes

### Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files

* Se han importado datos abiertos

#### Data transformation


In [8]:
tdata <- data |> st_transform(crs = 4326)

## <font color='green'>Data Save</font>

Este proceso, puede copiarse y repetirse en aquellas partes del notebbok que necesiten guardar datos.
Recuerde cambiar las cadenas añadida del fichero para diferenciarlas

<font color='tomato'> Identificamos los datos a guardar</font>

In [9]:
data_to_save <- tdata


<font color='tomato'>Estructura de nombre de archivos:</font>

* Código del caso de uso, por ejemplo "CU_04"
* Número del proceso que lo genera, por ejemplo "_05".
* Número de la tarea que lo genera, por ejemplo "_01"
* En caso de generarse varios ficheros en la misma tarea, llevarán _01 _02 ... después
* Nombre: identificativo de "properData", por ejemplo "_zonasgeo"
* Extensión del archivo

Ejemplo: "CU_04_05_01_01_zonasgeo.json, primer fichero que se genera en la tarea 01 del proceso 05 (Data Collection) para el caso de uso 04 (vacunas)

Importante mantener los guiones bajos antes de proceso, tarea, archivo y nombre

### Proceso 05

In [10]:
caso <- "CU_45"
proceso <- '_05'
tarea <- "_01"
archivo <- ""
proper <- "_municipios_geo"
extension <- ".json"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [ ]:
# file_save <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out <- paste0(oPath, file_save)
# write_csv(data_to_save, path_out)

# cat('File saved as: ')
# path_out

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [11]:
file_save <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out <- paste0(oPath, file_save)
st_write(obj = data_to_save, 
         dsn = path_out, 
         driver = "GeoJSON",
         delete_dsn = TRUE)

cat('File saved as: ')
path_out

Deleting source `Data/Output/CU_45_05_01_municipios_geo.json' failed
Writing layer `CU_45_05_01_municipios_geo' to data source 
  `Data/Output/CU_45_05_01_municipios_geo.json' using driver `GeoJSON'
Writing 179 features with 7 fields and geometry type Unknown (any).
File saved as: 

[1] "Data/Output/CU_45_05_01_municipios_geo.json"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [12]:
path_in <- paste0(iPath, file_save)
file.copy(path_out, path_in, overwrite = TRUE)

[1] TRUE

## <font color=' #2874a6 '>Main Conclusions</font>

<font color='steelblue'>List and describe the general conclusions of the analysis carried out​.</font>

### Prerequisites

Para que funcione este código se necesita:

* Las rutas de archivos `Data/Input` y `Data/Output` deben existir (relativas a la ruta del _notebook_)
* El paquete tcltk instalado para seleccionar archivos interactivamente. No se necesita en producción.
* Los paquetes sf, mapSpain, dplyr deben estar instalados.


### Configuration Management

This notebook has been tested with the following versions of R and packages. It cannot be assured that later versions work in the same way:
* R 4.2.2
* tcltk 4.2.2
* sf 1.0.9
* mapSpain 0.7.0
* dplyr 1.1.0


NOTA: En los resultados preliminares, problemas al instalar paquete geojsonio por dependencia con GDAL. Resuelta instalando todo el sistema gdal con homebrew. En todo caso para el piloto estamos solo trabajando con GEOJSON y no hace falta ese paquete.

### Data structures

#### Objeto `data`

- Hay 179 filas con información de las siguientes variables:
	* codauto
	* ine.ccaa.name
	* cpro
	* ine.prov.name
	* cmun
	* name
	* LAU_CODE
	* geometry


#### Observaciones generales sobre los datos

* Son datos oficiales

### Consideraciones para despliegue en piloto

* No se espera que esos datos cambien, se integrarán en el archivo a utilizar

### Consideraciones para despliegue en producción

* Se deben crear los procesos ETL en producción necesarios para que los datos de entrada estén actualizados

## <font color=' #2874a6 '>Main Actions</font>

#### <font color=' #2874a6 '>Acciones done</font>

<font color='steelblue'>Indicate the actions that have been carried out in this process</font>

- Se han obtenido los contornos de municipios
- Se ha guardado en archivo json

#### <font color=' #2874a6 '>Acctions to perform</font>

<font color='steelblue'>Indicate the actions that must be carried out in subsequent processes</font>

- Se deben unir estos datos al resto para el análisis

## <font color='RED'>CODE TO DEPLOY (PILOT)</font>

A continuación se incluirá el código que deba ser llevado a despliegue para producción, dado que se entiende efectúa operaciones necesarias sobre los datos en la ejecución del prototipo

<font color=' red '><b>Description</b></font>

- No hay nada que desplegar en el piloto, ya que estos datos son estáticos o en todo caso cambian con muy poca frecuencia, altamente improbable durante el proyecto.


<font color=' red '><b>CODE</b></font>

In [ ]:
# incluir código